sklearn/neighbors/_graph.py

```diff
@@ -36,6 +43,20 @@ def _query_include_self(X, include_self, mode):
     return X
 
 
+@validate_params(
+    {
+        "X": ["array-like", BallTree, KNeighborsMixin],
+        "n_neighbors": [Interval(Integral, 1, None, closed="left")],
+        "mode": [StrOptions({"connectivity", "distance"})],
+        "metric": "no_validation",  # metric is passed without param validation
+        # here, because VALID_METRICS are extracted and defined in
+        # :class:`~sklearn.neighbors._base.NeighborsBase`
+        "p": [Interval(Real, 1, None, closed="left")],
```

Comment:
@glemaitre There might be an issue here, or I'm on a wrong path, and I hope you can help.

`kneighbors_graph` and also `NeigborsBase` accept floats for p, as long as they're bigger than 1 (or maybe >=1).

Since 
```
        Interval(Real, 1, None, closed="left")
        Interval(Real, 0, None, closed="right")
```
are intervals of floats, I don't think they're both the same.



There's a file in the metrics that explains the possible values for `p:` `sklearn/metrics/_dist_metrics.pyx ` 
I wondered if line 1277 (the last one here) should be changed:

```
Minkowski Distance requires p >= 1 and finite. For p = infinity, use ChebyshevDistance. Note that for p=1, ManhattanDistance is more efficient, and for p=2, EuclideanDistance is more efficient.
    """
    def __init__(self, p, w=None):
        if p < 1:
            raise ValueError("p must be greater than 1")
```
Shouldn't it be `raise ValueError("p must be equal to or greater than 1")`? To account for what is  written there?


I don't understand why `None` should also be supported?
When I try it out and set `p = None`, then 
```
  File "/home/stefanie/Python/scikit-learn_dev/scikit-learn/sklearn/neighbors/_base.py", line 603, in _fit
    and self.effective_metric_params_["p"] < 1
TypeError: '<' not supported between instances of 'NoneType' and 'int'
```

Should I update the documentation with the information that p is of type `float`  and which values it can take? (Would also involve the docstring for `KNeighborsClassifier` and maybe other estimators, didn't check yet.) And for the test, I would pick values like p = 1, 0.5, 5.7 and float('inf')? (Please don't provide the code, I'd like to try this on my own, but I'd need a hint.)

Reply:
Actually you are reporting the same issue that here: https://github.com/scikit-learn/scikit-learn/issues/26548

As discussed in the issue, we should not raise an error and be lenient about `0 < p < 1`. With such `p` value, we don't have a true metric but it might still be useful for people to use it.

So we should tolerate value below 1 and we would expect the code then to raise a warning.

Reply:
I see, `Interval(Real, 0, None, closed="right") ` it is then. :) But what about the `None` value? I think it should not be accepted, because it raises a TypeError.

Reply:
Yep, it seems that we accepted something that we should not here. Let's `None` apart.
We might need to check that this `None` is not accepted in any place. I will open a PR, to check if we break some code.

Reply:
I open https://github.com/scikit-learn/scikit-learn/pull/26729 to check if we break something.

Reply:
OK, so the reason for the failure is that you can either pass `p` or set `p=None` and set `metrics_params={"p": 2}`. They are equivalent. So we should probably better handle the default and be consistent across estimators.

I would say that the best here is to accept `None` and we see in a subsequent PR what would be the strategy.

Reply:
Yes, I see. 

I've now also added a test for the error handling for the p values as you had proposed above, but I'm really not sure if this is what you had in mind.

In more general terms: What would be a good hierarchy of what needs to be tested for?  [Testing and improving test coverage](https://scikit-learn.org/dev/developers/contributing.html#testing-and-improving-test-coverage) is not very wordy here.